### DISCUSSION TITLE
motion state translation reset to (0,0,0) in some cases

### DISCUSSION THREAD

**Discussion by gjaegy** [bulletphysics/bullet3#3633] [General]
Hi,

just spent my day fixing some border-line case bug, where my motion state transform was being set to identity/(0,0,0) during the step update.

This is what happens:
- create and initialize a motion state with the current position/orientation
- create a new rigid body using that motion state
- before adding the rigid body to the world, call btRigidBody::setWorldTransform() to initialize the rigid body position to be safe
- add rigid body to world
- step the physic simulation (::stepSimulation()) with a timeStep < fixedTimeStep (for instance, 5ms for a fixedTimeStep of 16.66ms)

This will actually call btDiscreteDynamicsWorld::synchronizeSingleMotionState() for that body, and update the motionstate using the value returned by btRigidBody::getInterpolationWorldTransform().
However, since the simulation hasn't actually stepped, and because the rigid body has just been added, btRigidBody::getInterpolationWorldTransform() returns an identity transform, which override the transform stored in the  motion state.

My workaround is to call both btRigidBody::setWorldTransform() & btRigidBody::setInterpolationWorldTransform() before adding the rigid body to the world. But that is definitively counter-intuitive.

I hope that makes sense. If not don't hesitate to get back to me.

Cheers
Greg

**Comment by erwincoumans** [comment]
Thanks for the report. We probably should call setInterpolationWorldTransform when adding the body to the world.
